tensorboard/data/server/main.rs

```diff
@@ -15,35 +15,132 @@ limitations under the License.
+        res.runs.push(data::Run {
+            name: "train".to_string(),
+            start_time: 1605752017.0,
             ..Default::default()
-        };
-        Ok(Response::new(response))
+        });
+        Ok(Response::new(res))
+    }
+
+    async fn list_scalars(
+        &self,
+        _request: Request<data::ListScalarsRequest>,
+    ) -> Result<Response<data::ListScalarsResponse>, Status> {
+        let mut res: data::ListScalarsResponse = Default::default();
+        let mut run: data::list_scalars_response::RunEntry = Default::default();
+        run.run_name = "train".to_string();
+        run.tags.push(data::list_scalars_response::TagEntry {
+            tag_name: "accuracy".to_string(),
+            metadata: Some(data::ScalarMetadata {
+                max_step: 5,
+                max_wall_time: 1605752022.0,
+                ..Default::default()
+            }),
+            ..Default::default()
+        });
+        res.runs.push(run);
+        Ok(Response::new(res))
+    }
+
+    async fn read_scalars(
+        &self,
+        _request: Request<data::ReadScalarsRequest>,
+    ) -> Result<Response<data::ReadScalarsResponse>, Status> {
+        let mut res: data::ReadScalarsResponse = Default::default();
+        let mut run: data::read_scalars_response::RunEntry = Default::default();
+        run.run_name = "train".to_string();
+        run.tags.push(data::read_scalars_response::TagEntry {
+            tag_name: "accuracy".to_string(),
+            data: Some(data::ScalarData {
+                step: vec![0, 1, 2, 3, 4],
+                wall_time: (0..=4).map(|i| 1605752018.0 + (i as f64)).collect(),
+                value: vec![0.1, 0.5, 0.8, 0.9, 0.95],
+                ..Default::default()
+            }),
+            ..Default::default()
+        });
+        res.runs.push(run);
+        Ok(Response::new(res))
+    }
+
+    async fn list_tensors(
+        &self,
+        _request: Request<data::ListTensorsRequest>,
+    ) -> Result<Response<data::ListTensorsResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn read_tensors(
+        &self,
+        _request: Request<data::ReadTensorsRequest>,
+    ) -> Result<Response<data::ReadTensorsResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn list_blob_sequences(
+        &self,
+        _request: Request<data::ListBlobSequencesRequest>,
+    ) -> Result<Response<data::ListBlobSequencesResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn read_blob_sequences(
+        &self,
+        _request: Request<data::ReadBlobSequencesRequest>,
+    ) -> Result<Response<data::ReadBlobSequencesResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    type ReadBlobStream =
+        Pin<Box<dyn Stream<Item = Result<data::ReadBlobResponse, Status>> + Send + Sync + 'static>>;
```

Comment:
Flagging this type signature, since it’s imposing and will probably
raise questions…

This `Pin<T>` business mostly to do with async code. In any function,
you can hold a variable on your stack, take a reference to it, and use
that reference:

```rust
fn foo() -> (i32, i32, i32) {
    let mut xyz = (1, 2, 3);
    let x = &mut xyz.0;
    *x = 777;
    xyz  // returns (777, 2, 3)
}
```

But the defining characteristic of async functions is that they can be
suspended and resumed at `await` points. So what happens if there’s an
`await` point in there?

```rust
async fn foo(new_x: impl Future<Output = i32>) -> (i32, i32, i32) {
    let mut xyz = (1, 2, 3);
    let x = &mut xyz.0;
    *x = new_x.await;
    xyz
}
```

When we invoked `foo`, it runs until the `new_x.await` expression, then
suspends until the `new_x` future resolves. When it suspends, its stack
frame is recorded and stashed away. Once `new_x` resolves, the event
loop can resume the invocation of `foo` by restoring the contents of its
stack frame. But the stack frame may be at a different physical address
now, which means that while `xyz` is still valid (it’s just a 12-byte
block of memory with three integers in it), the reference `x` will not
be valid, because it pointed into the _old_ location of `xyz` on the old
stack frame.

Yet if you [try it out][play] you’ll see that the above function
compiles and runs just fine. How can this be?

The trick is that the “stack frame” used for the async function is
_pinned_. A [`Pin<&mut T>`][pin] is a mutable reference to `T` whose
value can never be moved until it is dropped, or unless it is
“move-safe”. A value is move-safe if it doesn’t have any references into
itself—so most types are move-safe, but the stack frame for our function
looks kind of like

```rust
struct FooFuture {
    xyz: (i32, i32, i32),
    x: &'self mut xyz,
}
```

where `&'self` is made-up syntax for “reference into self”. The
technical name for “move-safe” is [`Unpin`][unpin].

[pin]: https://doc.rust-lang.org/std/pin/struct.Pin.html
[unpin]: https://doc.rust-lang.org/std/marker/trait.Unpin.html

So, you can read the `ReadBlobStream` type definition as “a unique owned
pointer to a non-relocatable, dynamically dispatched stream, whose items
are either `ReadBlobResponse`s or `Status`es, where the stream itself
can be transferred across threads, can be shared among threads, and has
only static references (i.e., is permitted to live forever)”. A lot, to
be sure, but undeniably precise.

In practice, [actually creating the stream][s] is pretty straightforward
and doesn’t require any weird types.

[s]: https://github.com/wchargin/rustboard-server/blob/05550d7b356b94c7e0dea335030d4d11ab6510b8/src/main.rs#L468-L474

I hope that this helps describe the problem that pinning is meant to
solve. I haven’t gone into the details of how exactly it works. IMHO,
pinning is the most complicated corner of Rust, and I can’t claim to
understand it entirely. But if you have a vague understanding of the
concepts and how they relate to memory layout, then the compiler errors
and documentation are generally good enough to guide you to a solution
that works. I expect that over time I’ll come to really grok it more
deeply. :-)

[play]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a23a86217b4a28a1ef06c04762d1ce89